workflow/controller/controller.go

```diff
@@ -193,31 +199,81 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, podWorkers in
 	go wfc.wfInformer.Run(ctx.Done())
 	go wfc.wftmplInformer.Informer().Run(ctx.Done())
 	go wfc.podInformer.Run(ctx.Done())
-	go wfc.podLabeler(ctx.Done())
-	go wfc.podGarbageCollector(ctx.Done())
-	go wfc.workflowGarbageCollector(ctx.Done())
-	go wfc.archivedWorkflowGarbageCollector(ctx.Done())
 
-	go wfc.runTTLController(ctx)
-	go wfc.runCronController(ctx)
-	go wfc.metrics.RunServer(ctx)
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(ctx.Done(), wfc.wfInformer.HasSynced, wfc.wftmplInformer.Informer().HasSynced, wfc.podInformer.HasSynced) {
+		log.Fatal("Timed out waiting for caches to sync")
+	}
 
 	wfc.createClusterWorkflowTemplateInformer(ctx)
-	wfc.waitForCacheSync(ctx)
 
 	// Create Synchronization Manager
 	err := wfc.createSynchronizationManager()
 	if err != nil {
-		panic(err)
+		log.Fatal(err)
+	}
+
+	// HA support
+	if wfc.haEnabled {
```

Comment:
I think we can avoid this flag if I remember correctly. The leader election will work if there is only one replica of the app.

That way we can avoid creating two code paths and two sets of manifests that need testing and maintaining. I think by doing that we'll avoid future bugs.

Reply:
>I think we can avoid this flag if I remember correctly. The leader election will work if there is only one replica of the app.
Yes, it will work, the only thing we are adding unnecessary leaders election call flow, RBAC, and health check API call for NON-HA deployments.

> That way we can avoid creating two code paths and two sets of manifests that need testing and maintaining. I think by doing that we'll avoid future bugs.
 I would see the flag and two manifests will give a more clear understanding to users what they are deploying in their environment.

Reply:
I get that - I'm aware it is a trade-off - but I wonder if  the better option is to have it by default:

* You only need to scale up your deployment to get HA - no additional configuration.
* We don't need to maintain and test the third set of manifests. 

I'm thinking this is a v3 feature, so we have time to discuss.